fxa-rust-client/src/oauth.rs

```diff
@@ -0,0 +1,333 @@
+use errors::*;
+use regex::Regex;
+use url::Url;
+
+// These character ranges are from the OAuth RFC,
+// https://tools.ietf.org/html/rfc6749#section-3.3
+const VALID_SCOPE_VALUE: &str = r"^[\x21\x23-\x5B\x5D-\x7E]+$";
+const VALID_SHORT_NAME_VALUE: &str = r"^[a-zA-Z0-9_]+$";
+const VALID_FRAGMENT_VALUE: &str = r"^[a-zA-Z0-9_]+$";
+
+enum ScopeValue {
+    ShortScope(ShortScopeValue),
+    URLScope(URLScopeValue),
+}
+
+struct ShortScopeValue {
+    has_write: bool,
+    names: Vec<String>,
+}
+
+impl ShortScopeValue {
+    pub fn new(value: &str) -> Result<ShortScopeValue> {
+        if !Regex::new(VALID_SCOPE_VALUE).unwrap().is_match(value) {
```

Comment:
Over in fxa-email-service Phil has been using a nice pattern of statically creating the `Regex` object, since IIUC the overhead of compiling the regex into state-machine for matching is non-trivial:

  https://github.com/mozilla/fxa-email-service/blob/master/src/validate/mod.rs#L11

Not a blocker but might be worthwhile to use a similar technique here as well.